Fix mod-morph (#1412)
zmkfirmware/zmk: docs/docs/behaviors/mod-morph.md

```
mods = <(MOD_LGUI|MOD_LSFT|MOD_RGUI|MOD_RSFT)>;
```

### Advanced configuration

`keep-mods`

> When a modifier specified in `mods` is being held, it won't be sent along with the morphed keycode, unless it is also specified in `keep-mods`. By default, `keep-mods` equals `0`.

Reviewer: ```suggestion
When a modifier specified in `mods` is being held, it won't be sent along with the morphed keycode unless it is also specified in `keep-mods`. By default `keep-mods` equals `0`, which means no modifiers will be preserved.
```

> When a modifier specified in `mods` is being held, it won't be sent along with the morphed keycode if it is also part of `masked-mods`. To sent all modifiers along with the morphed keycode, set `masked-mods` to `<0>`. By default, `masked-mods` equals `mods`.

Reviewer 2: Doesn't this mean the default behavior for mod morph is changed as part of this PR?

Do we think there's any reason to be concerned by that?

Author: My reading was that most considered the old default a bug (see #683). Personally, I could go either way. But I thought that the proposed default may simplify life for the typical use-case of mod-morph?

Reviewer 2: Plenty of voices there that would seem to suggest a new default behavior is fine... Wonder how many folks use it today is as and it works well?

Author: Also, one can always enforce that a mod gets sent along by adding it to the binding as implicit mod. Hopefully that'll be intuitive enough so that most people won't even need to adjust the masked-mods property?

Reviewer 2: Yeah, I am leaning towards this being an "in the greater good" breaking change in behavior, given the fact the original versions sending along the mods was "odd" in the first place.

@joelspadin @Nicell @okke-formsma Thoughts?

Reviewer 3: > Plenty of voices there that would seem to suggest a new default behavior is fine... Wonder how many folks use it today is as and it works well?

I am currently using it on my keyboard: Works great (nice work @urob ). And in my opinion you should change the default behavior as it really is somewhat odd (you can always add modifiers via functions but can not remove them later on).

Reviewer 4: > Yeah, I am leaning towards this being an "in the greater good" breaking change in behavior, given the fact the original versions sending along the mods was "odd" in the first place.
> 
> @joelspadin @Nicell @okke-formsma Thoughts?

I am using it and it is working for my use-case of shift-morphing to have `,;`, `.?` and `/\` keys (instead of the regular qwerty `,<`, `.>` and `\|`).
I'm also of the opinion that masking the held mods by default is more sensible than having to mask them explicitly.

To nitpick further, `masked_mods` makes sense at the implementation level (in `hid.c`) but it feels counter-intuitive in the user-facing behavior. Exposing something like `keep-mods` for the behavior would make more sense in my opinion.
The doc could read:
> When a modifier specified in `mods` is being held it won't be sent along with the morphed keycode, unless it is also specified in `keep-mods`.

`keep-mods` would be empty by default which would be equivalent to `masked-mods`=`mods` by default

Reviewer 5: > Yeah, I am leaning towards this being an "in the greater good" breaking change in behavior, given the fact the original versions sending along the mods was "odd" in the first place.

I agree with everyone else here. By default mods should be masked. This was not done originally because Okke didn't have the time to look into it AFAIK. I think this is better behavior, and I can't imagine many people taking advantage of this odd default behavior.

Author: > To nitpick further, `masked_mods` makes sense at the implementation level (in `hid.c`) but it feels counter-intuitive in the user-facing behavior. Exposing something like `keep-mods` for the behavior would make more sense in my opinion.
> 
> The doc could read:
> 
> > When a modifier specified in `mods` is being held it won't be sent along with the morphed keycode, unless it is also specified in `keep-mods`.
> 
> `keep-mods` would be empty by default which would be equivalent to `masked-mods`=`mods` by default

Re: masked-mods vs keep-mods, there's a minor restriction associated with the latter in that one wouldn't be able to mask any mods *other* than the ones triggering the mod-morph.  Not sure how much that's a restriction in practice, but it may limit some creative uses?

Reviewer 6: Masking the mods was indeed something that I wanted to do but didn't make it into the original PR.

Masking by default seems to make sense. I don't have an opinion on the `masked-mods` vs `keep-mods` debate.

Reviewer 2: > > To nitpick further, `masked_mods` makes sense at the implementation level (in `hid.c`) but it feels counter-intuitive in the user-facing behavior. Exposing something like `keep-mods` for the behavior would make more sense in my opinion.
> > The doc could read:
> > > When a modifier specified in `mods` is being held it won't be sent along with the morphed keycode, unless it is also specified in `keep-mods`.
> > 
> > 
> > `keep-mods` would be empty by default which would be equivalent to `masked-mods`=`mods` by default
> 
> Re: masked-mods vs keep-mods, there's a minor restriction associated with the latter in that one wouldn't be able to mask any mods _other_ than the ones triggering the mod-morph. Not sure how much that's a restriction in practice, but it may limit some creative uses?

That's a good call out. I think potentially useful to support that for cases we're not imaging yet for this.

Reviewer 4: > > Re: masked-mods vs keep-mods, there's a minor restriction associated with the latter in that one wouldn't be able to mask any mods _other_ than the ones triggering the mod-morph. Not sure how much that's a restriction in practice, but it may limit some creative uses?
> 
> That's a good call out. I think potentially useful to support that for cases we're not imaging yet for this.

I hadn't thought that far. 
My issue was that `masked-mods` looked like an implementation detail leaking into the API for no good reason.
"Future use-cases" could be reason enough to go with it; and with the with the new default I assume it wouldn't show up in the vast majority of cases anyway.

Reviewer: I agree with snoyer, in addition if we are changing the default then it's also fair to break not-yet-imagined use cases that `masked-mods` might allow. Making it more intuitive for users should take priority here IMO.

Reviewer 2: Thinking about this more... I think I'm now leaning towards `keep-mods` for this. If we really need to filter other mods not specified in the morph later, we can solve that another way.

Author: The main use I could imagine is to set up a "dummy-mod-morph" which doesn't actually do anything other than masking mods. Currently, this is the only way to access the backend hid-functions `zmk_hid_masked_modifiers_set` and  `zmk_hid_masked_modifiers_clear` from the user side. But I agree that if there is demand for that, it would probably better to create a new "negative-mods" behavior that more directly exposes the masked_mods functionality.

I just pushed a new commit that switches from masked-mods to keep-mods. I tried out a couple of edge cases and it all works well on my side.

Reviewer 7: Having a behavior to set/clear masked mods would also be useful for sending strings, as you could lock out mods before sending the string and restore them after. That said, if the string is too long and it overflows the behavior queue, the restore behavior would never be sent and you wouldn't be able to use those mods anymore, so that wouldn't be great. Maybe instead we could add an optional mask property to specific behaviors, so for example you could add a `&kp` variant that masks mods and use that instead of `&kp` when sending strings.

```
mods = <(MOD_LGUI|MOD_LSFT|MOD_RGUI|MOD_RSFT)>;
```

### Advanced configuration

> `masked_mods`

Reviewer: I believe we prefer hyphenated property names by convention, so `masked-mods` would be the right format for the docs (and yaml/dtsi).

Author: done

When a modifier specified in `mods` is being held, it won't be sent along with the morphed keycode if it is also part of `masked_mods`. To sent all modifiers along with the morphed keycode, set `masked_mods` to `<0>`. By default, `masked_mods` equals `mods`.

For example, the following configuration morphs `LEFT_SHIFT` + `BACKSPACE` into `DELETE`, and morphs `RIGHT_SHIFT` + `BACKSPACE` into `RIGHT_SHIFT` + `DELETE`.

```
/ {
    behaviors {
        bspc_del: backspace_delete {
            compatible = "zmk,behavior-mod-morph";
            label = "BACKSPACE_DELETE";
            #binding-cells = <0>;
            bindings = <&kp BACKSPACE>, <&kp DELETE>;
            mods = <(MOD_LSFT|MOD_RSFT)>;
            masked_mods = <(MOD_LSFT)>;
        };
    };
};
```

### Limitations

> In some circumstance, when the morphed keycode is a hold-tap, the modifier gets sent along with the hold-tap regardless of the specification of `masked_mods`.

Reviewer: ```suggestion
Under certain circumstances where the morphed keycode is a hold-tap, the modifier gets sent along with the hold-tap regardless of the `masked_mods` value.
```

Author: Actually, that part is hopefully obsolete given b903e4f3710d0f331407c6ad26ce466a3a310845. Still waiting for @nickconway to confirm before removing the note.

Author: It seems the fix worked and mod-morphing into hold-tap no longer ignores masked-mods. So I am dropping the disclaimer
